3단계 - 질문 삭제하기 리팩터링 리뷰 요청 드립니다.

yomni opened the pull request:
안녕하세요!
우선 꼼꼼하게 리뷰해주셔서 감사합니다 ^_^

이번 단계에서는 리팩터링 미션 답게 여러 부분 고민하거나, 실제 변경해야 하는 부분도 많아서  
다소 다른 단계보다 난이도가 갑자기 확 올라간 느낌이었습니다!

미션 진행 시 몇가지 궁금한 점이 생겨 질문 남깁니다..!

Q1) ATDD(인수 테스트 주도 개발) 강의를 수강한 이후 이번 미션을 진행하다 보니,  
이번 미션의 요구사항이 마치 `인수 테스트의 시나리오`처럼 느껴졌습니다.(인수 테스트의 프리퀄(?)이라고나 할까요?)  
제가 알게 모르게 느꼈던 것처럼 이번 미션이 일종의 인수 테스트 라고 생각해봐도 되는 걸까요?

Q2) Answer와 Question의 중복되는 Attribute 에 대해 공통되는 로직이나 구현부분이 많아서  
BaseContentEntity 라는 별도의 클래스로 뽑아보았습니다. 

이 부분에서 가장 고민이 많았는데,,,  
`JPA의 상속(단일 테이블 전략)`을 사용할 지 vs `MappedSuperclass` 로 구현할 지
고민이 많이 되었습니다. 결국 MappedSuperclass로 구현 했는데,
리뷰어님께서는 동일한 고민을 하셨을 지,, 만약 리뷰어님께서 구현 하신다고 하면 어떤 방법으로 하는 것이 바람직하다고  
생각 하시는 지 궁금합니다!!

Q3) 세번째 질문도 구현에 관한 부분인데, List<Answer> --> Answers 라는 일급 컬렉션으로 변경하면서,
JPA의 @Embedded 타입을 사용하여 구현하였습니다. (값 타입) 이 부분도 바람직한 구현인 지 궁금합니다..!!

질문이 많아서 죄송합니다!!

3단계 피드백 미리 감사드립니다 😆

yomni committed:
docs: step2 피드백 관련 기능 목록 작성

yomni committed:
refactor: ManyToOne - FetchType.LAZY 적용

yomni committed:
refactor: 연관관계 편의 메서드 구현

yomni committed:
docs: 3단계 - 질문 삭제하기 리팩터링 구현 기능목록 작성

요구사항을 면밀히 분석하여 기능 목록 작성
> ATDD(인수 테스트 주도 개발) 강의를 듣고 난 후에 이 미션을 진행해보니,
일종의 인수테스트의 프리퀄(?)으로도 느껴지네요.

아래에 주요 기능 목록에 대해 정리
1. Content 신규 도메인 설계
- id, writer, deleted, content 속성을 가진 질문 / 답글의 상위 클래스
2. Service에 집중되어 있는 비즈니스 로직을 도메인으로 이전
- 삭제의 주요 로직 : 답글 -> 질문 순으로 삭제, 삭제된 이력을 `DeleteHistory`에 추가
이 로직을 각 도메인의 기능으로 분리하는 리펙터링 진행
3. 기능 목록에 작성하진 않았지만, List<Answer> answers 속성을 일급 컬렉션으로 분리하는 작업도
생각하고 있습니다.

yomni committed:
refactor: DefaultTime 사용을 @EnableJpaAuditing 으로 변경

yomni committed:
fix: DefaultTime MappedSuperclass 반영 누락부분 적용

yomni committed:
docs: MappedSuperclass와 상속을 동시에 사용할 수 있는 지 확인이 필요하여 기능목록 우선순위 변경

yomni committed:
refactor: Answer 도메인 기능 구현

- 삭제 시 writer가 로그인 사용자와 같은 지 검증 로직
- 삭제 시 deleted 값 삭제 상태로 변경
- 삭제 후 deleteHistory에 삭제된 데이터가 존재해야 함

yomni committed:
refactor: Question 기능 구현

- 삭제 시 deleted 값 삭제 상태로 변경
- 삭제 시 writer가 로그인 사용자와 같은 지 검증로직

yomni committed:
style: DefaultTime --> BaseDateTimeEntity로 이름 변경

yomni committed:
refactor: 연관관계정보 변경

**_User : DeleteHistory = 1 : 1 --> User : DeleteHistory = 1 : * 로 변경했습니다._**

yomni committed:
refactor: 일급 컬렉션 생성 List<Answer> answers --> Answers answers

yomni committed:
style: 테스트 메소드명 한글로 변경

yomni committed:
refactor: 일부 메소드 로직 분리

yomni committed:
refactor: 질문 삭제에 대한 리팩터링 기능 구현

1. 로그인 유저가 지우려는 질문의 유저인지 확인
2. 질문의 상태를 변경
3. 질문과 연관있는 답변들을 모두 삭제(로그인 유저와 동일한 경우)
4. 질문 + 답변 모두 DeleteHistory 테이블에 적재

yomni committed:
docs: 추가 리팩터링 기능 목록 작성

yomni committed:
refactor: List<DeleteHistory> deleteHistories를 일급 컬렉션으로 변경

yomni committed:
refactor: Content의 공통 속성을 BaseContentEntity로 MappedSuperclass로 분리

yomni committed:
refactor: 사용하지 않는 setWriter 삭제

wooobo commented on src/main/java/qna/domain/BaseContentEntity.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.Objects;
+
+@MappedSuperclass
+public abstract class BaseContentEntity extends BaseDateTimeEntity {
```
`BaseContent`라는 이름을 가졌지만, 
예상보다 많은 책임을 가지는 슈퍼클래스가 탄생한것 같아요.
그 만큼 변경에 취약 할 것이라 예상되어요(변경해야될 이유가 많기 때문에)

`BaseDateTimeEntity` 클래스는 이름에 걸맞게 `시간`에 대한 상태만 관리하는 부분이 저는 좀 더 안정되어 보이는것 같아요.

저는 `BaseContentEntity`가 다소 과한것 같은데 이부분 어떻게 생각하시나요? :thinking:

yomni replied:
**_중복 코드는 최대한 배제하자_** 라는 제 나름의 원칙에 의해 시작된 설계였습니다.

다시 생각해보니 객체의 역할과 책임의 관점에서 보면,   
- id : 해당 Entity가 id를 갖고 있는 것이 말씀하신 것처럼 바람직해 보입니다.
- writer : Question / Answer는 공동의 writer를 가지고 있지만, 각각의 Entity의 역할과 책임이 다른 만큼 각자 다른 기능으로 확장될 가능성이 높아보입니다.
- deleted : 이미 3단계의 요구사항에서 부터 각자 다른 기능을 요구하고 있으므로, 각자 가지고 있는 것이 바람직해 보입니다.

사실상 공통의 속성으로 `MappedSuperclass`로 뽑아낼 수 있는 부분이 content 속성인 것 같은데요,, 지금으로선 사실 굳이 공통 속성 Entity로 뽑아낼 만큼 가치가 있는지는 의문입니다.

결과적으론 아무리 공통된 속성이 있어도, 역할과 책임이란 측면에서 검토해보면 각각의 Entity가 공통된 속성을 각자 가지고 있는 것이 바람직한 것으로 보여지네요..!

깊은 고민을 할 수 있게 피드백 주셔서 감사합니다 👏🏻

wooobo commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+
+    public List<DeleteHistory> deleteHistories() {
```
`values` 라는 네이밍을 저는 생각을 해보았는데 어떻게 생각하시나요 :smile:

yomni replied:
values가 조금더 자연스러운 네이밍 같네요 :smile:

wooobo commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public DeleteHistories(List<DeleteHistory> deleteHistories) {
+        this.deleteHistories.addAll(deleteHistories);
```
```suggestion
        this.deleteHistories = deleteHistories;
```

위 코드가 더 깔끔할것 같아요!
개인적으로는 생성자에 `validate` 제외한 코드는 없는게 더 좋다라는 관점이에요
생성자는 순수한 그대로 작동되어야 안정된다는 생각때문이에요.

yomni replied:
생성자는 `validate` 제외한 코드가 없는게 더 좋다  
왜냐하면 생성자는 순수한 그대로 작동되어야 안정된다

공감가네요! 조언 리뷰 감사합니다!

wooobo commented on src/test/java/qna/service/QnaServiceTest.java:
```diff
@@ -46,7 +45,7 @@ public void setUp() throws Exception {
     public void delete_성공() throws Exception {
         when(questionRepository.findByIdAndDeletedFalse(question.getId())).thenReturn(Optional.of(question));
-        when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer));
+//        when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer));
```
사용하지 않는 코드는 지속적으로 제거하면서 관리하는게 더 좋다고 생각해요!

wooobo replied:
`diff` 코드에는 보이지 않지만, `delete_다른_사람이_쓴_글`  에 `throws Exception` 필요 없는듯해요

yomni replied:
앗 사용하지 않거나 주석처리된 코드는 삭제 하겠습니다!

wooobo commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,100 +1,60 @@
-    }
-
     public void addAnswer(Answer answer) {
         answer.toQuestion(this);
```
개인적으론 `toQuestion`이 필요할지 의문이에요.
이미 생성자를 통해 주입을 받고있어서요. :thinking:

yomni replied:
Answer가 기본 생성자의 접근제어자를 public으로 두었습니다.
물론, 기본 생성자로 인스턴스를 생성하는 경우는 드문 경우 일 것 같습니다.

하지만, addAnswer라는 `연관관계 편의 메서드`의 기능적인 측면에서는
toQuestion으로 확실하게 서로의 연관 관계를 맺어주는 것이 바람직해 보입니다.

안정적인 연관 관계 매핑을 위해, toQuestion 안에 `Answer`가 해당 `Question`에 달릴 수 있는 상태인지 확인하는 로직을 추가해보았습니다. 이런 방어로직 추가는 어떤 지 궁금합니다!

wooobo replied:
네! 용민님 생각이 궁금했는데, 잘 설명해주셔서 감사합니다!
충분히 고민하신거라면 이견없습니다 :smile:

wooobo commented on README.md:
```diff
@@ -59,8 +59,66 @@ alter table question
     foreign key (writer_id)
```
>Q1) ATDD(인수 테스트 주도 개발) 강의를 수강한 이후 이번 미션을 진행하다 보니,
이번 미션의 요구사항이 마치 인수 테스트의 시나리오처럼 느껴졌습니다.(인수 테스트의 프리퀄(?)이라고나 할까요?)
제가 알게 모르게 느꼈던 것처럼 이번 미션이 일종의 인수 테스트 라고 생각해봐도 되는 걸까요?

A) 저의 생각으로는, 개념 적인것 같아요. 그래서 무엇은 무엇이다라고 딱 정의 내리기가 애매한것 같다는 생각이에요.
`인수 테스트`는 `도메인테스트`가 아닌, 실제 어플리케이션의 사용자와의 상호작용 접점부분에 대한 시나리오 테스트가 아닌가 싶어요(이외 에도 다양한 부분이 있을수 있습니다)
요구 사항인 `답변이 없는 경우 삭제가 가능하다.`을 예로들어보겠습니다.

도메인에 대한 비지니스 규칙이므로 현재의 미션에서는 `단위테스트`개념이 맞지 않나 생각되어요. 
왜냐면, `DB` 라던지 `Service`라던지 그런 부분을 제외하고 `Question`이 가져야만 하는 규칙이라고 생각하기 때문이에요. 

물론, `인수 테스트` 시나리오에 `답변이 없는 경우 삭제가 가능하다.` 가 있을 수는 있겠죠 ^^

저의 생각이므로 참고 해주시고, 다양한 블로그를 찾아보시면 더 좋을것 같아요!

>Q2) Answer와 Question의 중복되는 Attribute 에 대해 공통되는 로직이나 구현부분이 많아서
BaseContentEntity 라는 별도의 클래스로 뽑아보았습니다.
이 부분에서 가장 고민이 많았는데,,,
JPA의 상속(단일 테이블 전략)을 사용할 지 vs MappedSuperclass 로 구현할 지
고민이 많이 되었습니다. 결국 MappedSuperclass로 구현 했는데,
리뷰어님께서는 동일한 고민을 하셨을 지,, 만약 리뷰어님께서 구현 하신다고 하면 어떤 방법으로 하는 것이 바람직하다고
생각 하시는 지 궁금합니다!!

A) 네! 맞아요. `ID` 같은거 계속 반복되는데... 분리하면 어떨까 생각되긴해요.
현재는 `BaseContentEntity`가 다소 많은 메시지를 받고 있지 않는 생각이 들어요.
저 같은 경우에는 같은 행동을 하는 상태끼리 묶거나 할것 같아요. `BaseDateTimeEntity` 처럼요.

추가적으로 고민해보면...
`ID` 같은 컬럼은 `BaseID` 이렇게 `ID`의 컬럼만을 가지는 클래스로 분리시킬것 같아요.
왜냐면, `BaseContentEntity` 상속받는 테이블중 `ID`컬럼 이름이 변경된다면 여러 이슈가 있겠죠 ^^

---

[리팩토링 - 디메테르의 법칙(The Law of Demeter),최소 지식 원칙](https://coding-start.tistory.com/264)

>Q3) 세번째 질문도 구현에 관한 부분인데, List --> Answers 라는 일급 컬렉션으로 변경하면서,
JPA의 @Embedded 타입을 사용하여 구현하였습니다. (값 타입) 이 부분도 바람직한 구현인 지 궁금합니다..!!

A) 
저는 괜찮은것 같아요~
왜냐하면, `Answers` 에게만 던질 메시지가 존재하기 때문이에요
`Answers.delete(User loginUser)` 처럼요.

yomni replied:
상세하게 답변 주셔서 감사합니다.

특히 많은 고민을 했던 `BaseContentEntity`의 경우 저는 **_중복코드의 최소화_** 라는 관점에서 본 반면,  
리뷰어님께서는 _**같은 행동이나 같은 상태로 묶일 수 있는 대상인지**_ 로 판단하시는 점에서 더욱 더 객체지향적으로 생각 하신 것 같아 답변만으로도 제가 많이 배운 것 같습니다 🙌🏻

감사합니다!

wooobo commented on src/test/java/qna/domain/UserTest.java:
```diff
@@ -18,8 +19,9 @@ public class UserTest {
+    void 유저저장_성공() {
         userRepository.save(JAVAJIGI);
         userRepository.save(SANJIGI);
         flushAndClear();
```
`AssertTrue(javajigi.isSame(JAVAJIGI))`
개체 비교하는 패턴으로 한다면, `getter`를 사용안해도 될것 같네요!

yomni replied:
아..! 객체에게 묻는 연습이 아직 부족한거같네요 😓 감사합니다!

wooobo reviewed (changes requested):
안녕하세요~ 3단계에서 정말 많은 고민을 해보셨네요! :100: 

질문에 대한 저의 생각과 코멘트 남겨 드렸는데 확인 후 재요청 부탁드릴께요~

yomni committed:
docs: 3단계 피드백 개선점 리스트업

yomni committed:
refactor: 주석 코드 or 사용되지 않는 부분 코드 삭제

yomni committed:
refactor: deleteHistories() --> values() 자연스러운 네이밍으로 변경

yomni committed:
refactor: User에서 isSame() 개체 비교 메서드 정의

yomni committed:
refactor: DeleteHistories의 순수한 생성자로 변경

yomni committed:
refactor: BaseContentEntity 삭제

BaseContentEntity의 필요성에 대한 고찰을 해본 결과, 각각의 속성에 따라 역할이 다르게 확장될 가능성이 있는 것들은 각자의 Entity로 다시 복원시켰고, Content는 분리하는 의미가 다소 부족해서 결과적으론 BaseContentEntity를 삭제하는 것으로 변경하였습니다.

yomni committed:
refactor: Answer.toQuestiond의 유효성 검증에 대한 방어로직 추가

yomni committed:
docs: 피드백 구현 부분 반영

wooobo commented on src/main/java/qna/domain/Question.java:
```diff
@@ -45,46 +51,45 @@ public boolean isOwner(User writer) {
 
-    public boolean isDeleted() {
-        return deleted;
+    private void setDeleted(boolean deleted) {
```
```
private void deleted() {
  this.deleted = true;
}
```

`setter` 보다는 좀 더 명확하게 메시지를 던져 볼수 있겠네요!

wooobo commented on README.md:
```diff
@@ -121,4 +121,12 @@ alter table question
+- [x] Question과 Answer의 공통 부분...
+
+### 3단계 피드백
+- [ ] BaseContentEntity 에 대한 과도한 책임 부여 --> Content에 집중된 책임과 역할에 대해 생각해 볼 것
```
정리 좋네요 :+1:

wooobo reviewed (approved):
3단계 피드백까지 잘정리해주시고 마무리 잘 해주셨네요 :+1: 

몇가지 코멘트 남겨드렸는데, 확인 부탁드려요~

이만 `merge`진행해도 될것같습니다!
JPA 미션진행하느라 수고많으셨습니다!
남은 미션 응원하겠습니다.
화이팅입니다~
